Add a method for dumping profiles as self-contained HTML w/ timeline visualization

fitzgen opened the pull request:
This commit adds the `--profiler-trace-path` flag. When combined with `-p` to
enable profiling, it dumps a profile as a self-contained HTML file to the given
path. The profile visualizes the traced operations as a Gantt-chart style
timeline.

Example output HTML file: http://media.fitzgeraldnick.com/dumping-grounds/trace-reddit.html

Mostly I made this because I wanted to see what kind of data the profiler has, and thought that this might be useful for others as well. I'm happy to add tests if we can figure out how to integrate them into CI, but I'm not sure how that would work.

Thoughts?

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/10694)

<!-- Reviewable:end -->

fitzgen committed:
Add a method for dumping self-contained HTML timeline profiles

This commit adds the `--profiler-trace-path` flag. When combined with `-p` to
enable profiling, it dumps a profile as a self-contained HTML file to the given
path. The profile visualizes the traced operations as a gant-chart style
timeline.

fitzgen commented:
Here is a screenshot of a generated timeline: http://i.imgur.com/OhWsm8m.png

fitzgen commented:
I guess it isn't immediately obvious, but you can zoom in and out with the mouse wheel / trackpad by focusing the upper slider part and scrolling. You can also drag the view left and right via the upper thing as well.

metajack commented:
cc @nox @pcwalton 

This is really cool! The reddit trace looks a little scary tough. Were you clicking around or something? Why is it running layout so many times?

@nox is trying to get hte Firefox tool to do this same thing working as well.

samlh commented on components/profile/trace-dump.js:
```diff
@@ -0,0 +1,504 @@
+
+    // Remove idle time between traces. It isn't useful to see and makes
+    // visualizing the data more difficult.
+    removeIdleTime() {
```
It would be nice if this would leave a marker so you can tell where the breaks are. Also, it may be usefull to be able to toggle this on or off.

fitzgen replied:
> It would be nice if this would leave a marker so you can tell where the breaks are. Also, it may be useful to be able to toggle this on or off.

I found that the vast majority of time was idle, presumably waiting on the network, which is not yet traced by the profiler, and it was sort of hard to find interesting things. This could just as easily be interpreted as reason to trace more things :) We could definitely make it toggle-able or insert placeholders to show that there was idle time squashed, but I'd prefer to get the initial thing landed before futzing about with details too much.

fitzgen commented:
> The reddit trace looks a little scary tough. Were you clicking around or something?

Nope, just loaded it, let it settle down, and then closed it.

> Why is it running layout so many times?

No idea :) The first one to turn on the lights is the first one to see the mess ;)

fitzgen commented:
Ok so the tidy script reports a couple of unsorted imports, I'm fixing those. But the tidy script doesn't seem to like the raw string literals containing html snippets:

```
$ ./mach test-tidy --no-progress
Checking files for tidiness...
./components/profile/trace_dump.rs:80: missing space after /
./components/profile/trace_dump.rs:81: missing space after /
./components/profile/trace_dump.rs:91: missing space after /
./components/profile/trace_dump.rs:92: missing space before =>
./components/profile/trace_dump.rs:99: missing space after /
./components/profile/trace_dump.rs:100: missing space after /
./components/profile/trace_dump.rs:101: missing space after /
```

I would prefer not to patch the tidy script, if possible. @edunham says that there isn't a way to tell the tidy script to ignore something either:

```
3:23 PM <qedunham> fitzgen: you've found the bug that https://github.com/servo/servo/blob/master/python/tidy/servo_tidy/tidy.py#L331 misses multiline strings
3:23 PM <qedunham> and I don't think there's a tidy ignore-this-bit syntax
```

Given these things, I am going to pull the HTML snippets out into their own files and `include_str!` them.

fitzgen commented:
Pushed a new version which should pass the tidy checks.

fitzgen commented:
Looks like there are cargo errors with conflicting serde_macros dependencies in the CI builds, but I haven't been able to reproduce that locally. I think I need to just bump all the `serde_macros` dependencies to 0.7.2?

```
error: failed to select a version for `serde_macros` (required by `profile`):
all possible versions conflict with previously selected versions of `serde_macros`
  version 0.7.2 in use by serde_macros v0.7.2
  possible versions to select: 0.7.0
```

jdm commented:
You'll need to run `./mach cargo-update -p profile` to update all the lock files appropriately.

fitzgen commented:
> You'll need to run ./mach cargo-update -p profile to update all the lock files appropriately.

Thanks! Ran that command and updated the pull request. Let's see if the CI is happy now...

fitzgen commented:
> You'll need to run ./mach cargo-update -p profile to update all the lock files appropriately.

@jdm, it seems that this did not fix the issue and it is still occurring. Any ideas?

jdm commented:
`./mach cargo-update -p serde_macros --precise 0.7.2` should do it.

fitzgen commented:
@jdm, sorry to keep bothering you, but...

It looks like the AppVeyor build broke while compiling `styles` and then there was a python error (in mach? or something else?) and I can't reproduce this locally. Any idea what's up? How I should proceed? What to dig into?

Thanks!

jdm commented:
That was fixed by #10749 and is unrelated to your changes.

jdm commented:
Whoops, I meant #10791 which is still in the queue.

fitzgen commented:
> Whoops, I meant #10791 which is still in the queue.

Great! So what's the next step for this pull request now that that PR merged? Does this need another try run?

jdm commented:
No, it's fine as it is.

SimonSapin commented:
Random notes from playing with the example output:

* The yellow overlay is "above" the scrollbar which prevents me from click-dragging it to scroll quickly.
* Zooming by scrolling on the upper slider should be 10~100 times faster. Right now scrolling the full length of my trackpad or my finger on a mouse wheel make the zoom level barely change.

---

Reviewed 6 of 10 files at r1, 6 of 6 files at r2, 4 of 4 files at r4.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[components/profile/Cargo.toml, line 33 \[r4\]](https://reviewable.io:443/reviews/servo/servo/10694#-KGHRygA9VIM8-rOxNb-:-KGHRygA9VIM8-rOxNb0:-200017539) ([raw file](https://github.com/servo/servo/blob/aa2cb546da7f9c8b6a3d16cbbb31b0dcf773e4eb/components/profile/Cargo.toml#L33)):*
Why moving sections around in this file? Empty sections are not necessary.

---

*[components/profile/Cargo.toml, line 46 \[r4\]](https://reviewable.io:443/reviews/servo/servo/10694#-KGHSDSP6IeIGzOl89e0:-KGHSDSQiZ469qs80vQh:-1527910609) ([raw file](https://github.com/servo/servo/blob/aa2cb546da7f9c8b6a3d16cbbb31b0dcf773e4eb/components/profile/Cargo.toml#L46)):*
This section and the previous (non-empty) one can be merged into one `[target.'cfg(target_os = "macos")'.dependencies]` section.

http://doc.crates.io/manifest.html#the-dependencies-section
https://doc.rust-lang.org/reference.html#conditional-compilation

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/10694#-:-KGHReSxA6P_NG70kOT1:1964680817)*
<!-- Sent from Reviewable.io -->

fitzgen commented:
> The yellow overlay is "above" the scrollbar which prevents me from click-dragging it to scroll quickly.

Do you mean the scrollbar on the right, for scrolling up and down? Would `overflow: auto` -> `overflow: scroll` fix it?

> Zooming by scrolling on the upper slider should be 10~100 times faster. Right now scrolling the full length of my trackpad or my finger on a mouse wheel make the zoom level barely change.

Interesting! What OS/hardware are you on? On OSX/MBPr, I can zoom all the way out in one swipe.

These kinds of little tweaks feel like the kinds of things we can handle in follow ups, to me -- do you feel the same?

---

Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[components/profile/Cargo.toml, line 33 \[r4\]](https://reviewable.io:443/reviews/servo/servo/10694#-KGHRygA9VIM8-rOxNb-:-KGIDlwu7BbuCuVH2cKm:-718544411) ([raw file](https://github.com/servo/servo/blob/aa2cb546da7f9c8b6a3d16cbbb31b0dcf773e4eb/components/profile/Cargo.toml#L33)):*
I think this happened when I did `cargo add serde_json`... I'll try and clean it up.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/10694#-:-KGIE7qVpRRwsc_jqm-7:-1048397591)*
<!-- Sent from Reviewable.io -->

SimonSapin commented:
Yes, that scrollbar. Tring to grab and move its handle selects text instead. Here is what it looks like:

![screenshot from 2016-04-26 19-09-44](https://cloud.githubusercontent.com/assets/291359/14827425/94823f6a-0be2-11e6-8264-49f986248f11.png)

I’m on Firefox / GNOME 3 with a Thinkpad. Scrolling with either the built-in trackpad or an external mouse gives similar result.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #10857) made this pull request unmergeable. Please resolve the merge conflicts.

fitzgen commented:
@SimonSapin: Ok, just rebased and force pushed.

* Zooming via mouse wheel or trackpad is 10x faster (a bit fast for me on OSX where scrolling has velocity and acceleration, but seems reasonable on jimb's Fedora/GNOME 3 where scrolling doesn't have that).

* Shifted the hover information over to make room for grabbing the scrollbar.

* Cleaned up the `Cargo.toml` messiness.

Cheers!

fitzgen commented:
Updated http://media.fitzgeraldnick.com/dumping-grounds/trace-reddit.html with a new trace of reddit with the new changes.

SimonSapin commented:
Travis-CI shows that a test need to be updated:

```rust
/home/travis/build/servo/servo/tests/unit/profile/time.rs:10:16: 10:44 error: this function takes 2 parameters but 1 parameter was supplied [E0061]
/home/travis/build/servo/servo/tests/unit/profile/time.rs:10     let chan = time::Profiler::create(None);

                                                                            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

r=me with that fixed

---

Reviewed 7 of 7 files at r5.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[components/profile/Cargo.toml, line 33 \[r4\]](https://reviewable.io:443/reviews/servo/servo/10694#-KGHRygA9VIM8-rOxNb-:-KGPKu8p1Sdpc2t3kAsQ:-862091122) ([raw file](https://github.com/servo/servo/blob/aa2cb546da7f9c8b6a3d16cbbb31b0dcf773e4eb/components/profile/Cargo.toml#L33)):*
Ok, I see. I don’t think `cargo add`is ready to be used at all yet: https://github.com/killercup/cargo-edit/issues/15

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/10694#-:-KGPM1wIlDL6inT_IGGS:-1823136252)*
<!-- Sent from Reviewable.io -->

fitzgen commented:
Fixed test.

SimonSapin commented:
@bors-servo r+

---

Reviewed 1 of 1 files at r6.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/10694#-:-KGRFThx8tOwOMMiSvz3:1236647940)*
<!-- Sent from Reviewable.io -->

bors-servo commented:
:pushpin: Commit 9fbb5c7 has been approved by `SimonSapin`

<!-- @bors-servo r=SimonSapin 9fbb5c720eb89947720fa745aa08fa4fed7143b4 -->

bors-servo commented:
:hourglass: Testing commit 9fbb5c7 with merge b8e2fa5...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/579), [arm32](http://build.servo.org/builders/arm32/builds/614), [arm64](http://build.servo.org/builders/arm64/builds/599), [linux-dev](http://build.servo.org/builders/linux-dev/builds/610), [linux-rel](http://build.servo.org/builders/linux-rel/builds/658), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/626), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/608), [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/636)

rzambre commented:
@fitzgen has this somehow broken? I'm seeing zeros instead of category-labels in the trace and it is all one color. Not sure which commit broke it.

fitzgen commented:
@rzambre not that I know of, although its been maybe a month since I last did it. One thing is that if the Servo process doesn't exit cleanly, the generated HTML can be incomplete. As for your zeros, I don't know... Maybe do a git log on the profiler component and see if anything looks like it changed the tracing data structures and didn't also change the JS code?
